presto-parquet/src/main/java/com/facebook/presto/parquet/readerv2/decoders/rle/BaseRLEBitPackedDecoder.java

```diff
@@ -0,0 +1,120 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.parquet.readerv2.decoders.rle;
+
+import org.apache.parquet.Preconditions;
+import org.apache.parquet.bytes.BytesUtils;
+import org.apache.parquet.column.values.bitpacking.BytePacker;
+import org.apache.parquet.column.values.bitpacking.Packer;
+import org.apache.parquet.io.ParquetDecodingException;
+
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+
+import static com.facebook.presto.parquet.readerv2.decoders.rle.BaseRLEBitPackedDecoder.Mode.PACKED;
+import static com.facebook.presto.parquet.readerv2.decoders.rle.BaseRLEBitPackedDecoder.Mode.RLE;
+
+public abstract class BaseRLEBitPackedDecoder
+{
+    private final boolean rleOnlyMode;
+
+    private final int bitWidth;
+    private final BytePacker packer;
+    private final InputStream in;
+
+    protected Mode mode;
+    protected int currentCount;
+    protected int currentValue;
+    protected int[] currentBuffer;
+
+    public BaseRLEBitPackedDecoder(int valueCount, int bitWidth, InputStream in)
+    {
+        Preconditions.checkArgument(bitWidth >= 0 && bitWidth <= 32, "bitWidth must be >= 0 and <= 32");
+        this.bitWidth = bitWidth;
+        if (bitWidth != 0) {
+            this.packer = Packer.LITTLE_ENDIAN.newBytePacker(bitWidth);
+            this.in = in;
+            this.rleOnlyMode = false;
+        }
+        else {
+            this.rleOnlyMode = true;
+            this.packer = null;
+            this.in = null;
+
+            this.mode = RLE;
+            this.currentValue = 0;
+            this.currentCount = valueCount;
+        }
+    }
+
+    public BaseRLEBitPackedDecoder(int rleValue, int rleValueCount)
+    {
+        this.rleOnlyMode = true;
+        this.bitWidth = 0;
+        this.packer = null;
+        this.in = null;
+
+        this.mode = RLE;
+        this.currentValue = rleValue;
+        this.currentCount = rleValueCount;
+    }
+
+    protected boolean readNext()
+            throws IOException
+    {
+        if (rleOnlyMode) {
+            // for RLE only mode there is nothing more to read
+            return false;
+        }
+
+        if (in.available() <= 0) {
+            this.currentCount = 0;
+            return false;
+        }
+
+        int header = BytesUtils.readUnsignedVarInt(this.in);
+        this.mode = (header & 1) == 0 ? RLE : PACKED;
+        switch (this.mode) {
+            case RLE:
+                this.currentCount = header >>> 1;
+                this.currentValue = BytesUtils.readIntLittleEndianPaddedOnBitWidth(this.in, this.bitWidth);
```

Comment:
static import BytesUtils.readIntLittleEndianPaddedOnBitWidth
do not use this to reference data members